internal/models/one_time_token.go

```diff
@@ -0,0 +1,330 @@
+		panic("at most 2 token types are accepted")
+	}
+
+	if err := query.First(oneTimeToken); err != nil {
+		if errors.Cause(err) == sql.ErrNoRows {
+			return nil, OneTimeTokenNotFoundError{}
+		}
+
+		return nil, errors.Wrap(err, "error finding one time token")
+	}
+
+	return oneTimeToken, nil
+}
+
+// FindUserByConfirmationToken finds users with the matching confirmation token.
+func FindUserByConfirmationOrRecoveryToken(tx *storage.Connection, token string) (*User, error) {
+	ott, err := FindOneTimeToken(tx, token, ConfirmationToken, RecoveryToken)
+	if err != nil && !IsNotFoundError(err) {
+		return nil, err
+	}
+
+	if ott == nil {
+		user, err := findUser(tx, "(confirmation_token = ? or recovery_token = ?) and is_sso_user = false", token, token)
+		if err != nil {
+			if IsNotFoundError(err) {
+				return nil, ConfirmationOrRecoveryTokenNotFoundError{}
+			} else {
+				return nil, err
+			}
+		}
+
+		return user, nil
+	}
+
+	return FindUserByID(tx, ott.UserID)
+}
+
+// FindUserByConfirmationToken finds users with the matching confirmation token.
+func FindUserByConfirmationToken(tx *storage.Connection, token string) (*User, error) {
+	ott, err := FindOneTimeToken(tx, token, ConfirmationToken)
+	if err != nil && !IsNotFoundError(err) {
+		return nil, err
+	}
+
+	if ott == nil {
+		user, err := findUser(tx, "confirmation_token = ? and is_sso_user = false", token)
+		if err != nil {
+			if IsNotFoundError(err) {
+				return nil, ConfirmationTokenNotFoundError{}
+			} else {
+				return nil, err
+			}
+		}
+
+		return user, nil
+	}
+
+	return FindUserByID(tx, ott.UserID)
+}
+
+// FindUserByRecoveryToken finds a user with the matching recovery token.
+func FindUserByRecoveryToken(tx *storage.Connection, token string) (*User, error) {
+	ott, err := FindOneTimeToken(tx, token, RecoveryToken)
+	if err != nil && !IsNotFoundError(err) {
+		return nil, err
+	}
+
+	if ott == nil {
+		return findUser(tx, "recovery_token = ? and is_sso_user = false", token)
+	}
+
+	return FindUserByID(tx, ott.UserID)
+}
+
+// FindUserByEmailChangeToken finds a user with the matching email change token.
+func FindUserByEmailChangeToken(tx *storage.Connection, token string) (*User, error) {
+	ott, err := FindOneTimeToken(tx, token, EmailChangeTokenCurrent, EmailChangeTokenNew)
+	if err != nil && !IsNotFoundError(err) {
+		return nil, err
+	}
+
+	if ott == nil {
+		return findUser(tx, "is_sso_user = false and (email_change_token_current = ? or email_change_token_new = ?)", token, token)
+	}
+
+	return FindUserByID(tx, ott.UserID)
+}
+
+// FindUserByEmailChangeCurrentAndAudience finds a user with the matching email change and audience.
+func FindUserByEmailChangeCurrentAndAudience(tx *storage.Connection, email, token, aud string) (*User, error) {
+	ott, err := FindOneTimeToken(tx, token, EmailChangeTokenCurrent)
+	if err != nil && !IsNotFoundError(err) {
+		return nil, err
+	}
+
+	if ott == nil {
+		ott, err = FindOneTimeToken(tx, "pkce_"+token, EmailChangeTokenCurrent)
```

Comment:
Do we need to consider One Time Tokens created using the implicit flow? GenerateLink doesn't support PKCE so existing links might not have a prefix

Reply:
Yes because that's what the existing logic does.

Reply:
Yup, and looks like it is [considered below](https://github.com/supabase/auth/pull/1558/files/2037c1f83319da7f237fb00c6bb0287d686b639a#diff-55a81f8dc53ee75ff592eec9c5e1fcb665f653d7593de7e9c8b8dadf7159fe83R269). If it doesn't find a `pkce_` one time token it will attempt to find a non-one-time-token with both pkce nad non-pkce prefix.


Thanks!